hypothesis-python/src/hypothesis/configuration.py

```diff
@@ -21,11 +25,50 @@ def set_hypothesis_home_dir(directory):
     __hypothesis_home_directory = None if directory is None else Path(directory)
 
 
-def storage_directory(*names):
+def storage_directory(*names, intent_to_write=True):
+    if intent_to_write and sideeffect_should_warn():
+        warnings.warn(
+            "Accessing the storage directory during import or initialization is "
+            "discouraged, as it may cause the .hypothesis directory to be created "
+            "even if hypothesis is not actually used. Typically, the fix will be "
+            "to defer initialization of strategies.",
+            HypothesisSideeffectWarning,
+            stacklevel=2,
+        )
+
     global __hypothesis_home_directory
     if not __hypothesis_home_directory:
         if where := os.getenv("HYPOTHESIS_STORAGE_DIRECTORY"):
             __hypothesis_home_directory = Path(where)
     if not __hypothesis_home_directory:
         __hypothesis_home_directory = __hypothesis_home_directory_default
     return __hypothesis_home_directory.joinpath(*names)
+
+
+def _sideeffect_never_warn():
+    return False
+
+
+if os.environ.get("HYPOTHESIS_WARN_SIDEEFFECT"):
+
+    def sideeffect_should_warn():
+        return True
```

Comment:
Why would we want to always warn?

Reply:
That's a bit of a long story, so I'll explain. Pardon the verbosity! I agree there's nothing pytest-specific about the check for plugins that we load ourselves, but pytest plugins (and conftest) are an instance of a wider check: Side effects during initialization, where initialization is not just the initial import.

We can't warn directly about side effects during post-import initialization, but we can _detect_ them. Even at no additional cost, since we can check whether the patching has been done - which happens on the first side-effect-inducing operation after import. And then the pytest plugin (or other test runners) can quite easily check if this is the case at start of session.

Now, it can warn that it has happened, but it can't see who did it or where, at least not easily [*]. That's where this setting comes in:
```
                "initialization, possibly causing slowdown or creation of files. To pinpoint and explain "
                "the problem, execute with environment 'PYTHONWARNINGS=error HYPOTHESIS_WARN_SIDEEFFECT=1'",
```
...which will cause the warning to happen, even after import is finished, and show the relevant stack trace.

[*] Yeah, we could record the stack of the first such call, but that's more effort than it's worth IMO

Reply:
> [*] Yeah, we could record the stack of the first such call, but that's more effort than it's worth IMO

Hm, or maybe. I guess it's easy enough to stash the stacktrace and message. Would simplify the messaging.

Reply:
Right, there are basically two ways to go about this:

1. Warn at the end of initialization, if we can see that a side-effectful operation happened.  
2. Warn on a side-effectful operation, if we're not yet finished initializing.  

I generally prefer (2), because the stack trace from `-Werror` makes it really easy to tell when this is happening.  

I guess the problem is how to tell whether we're in post-import initialization, but I think there we could set a global from the earliest possible pytest setup hook, and then unset it at the end?  We might still miss a few cases due to the ordering of hooks between our and other pytest plugins, but I think that applies equally to (1) for side effects triggered by pytest plugins.

At that point I suppose there's still a case for (1) to warn if we detect that there was a side-effect *between* finishing `import hypothesis` and delivery of the first pytest hook... but that seems concerningly prone to false alarms (c.f. `testdir.runpytest_inprocess()` test changes!).  Let's just skip detection in this case then?

Reply:
I agree (2) would be best, and we could do that as soon as possible (I was thinking even plugin-module execution, not wait for the import hook). I haven't done this because I am concerned that there is a significant chance of something happening during that hole in coverage. But hey, maybe it's not that important.

Reply:
Yeah... if there is a hole, we must check and deal with that anyway to revert the patching, so we may as well warn in that case, too. I'll follow this path then.